### ISSUE TITLE
unexpected state transition when backup node is rebooted, even though nopreempt is set

### ISSUE THREAD

**Issue by wsnipex** [acassen/keepalived#2209]
**Describe the bug**
master server A, backup server B (both configured as initial state backup, same priority and nopreempt)
If you reboot backup server B it will take over as master
This adds an unwanted transition

Note: the state transition only happens on *reboot* of the backup server, not when just keepalived is stopped / started

**To Reproduce**
Any steps necessary to reproduce the behaviour:

**Expected behavior**
rebooting a backup server should not take over as master, it should remain backup

**Keepalived version**
```
Keepalived v2.1.5 (07/13,2020)

Copyright(C) 2001-2020 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 4.18.0
Running on Linux 4.18.0-372.26.1.el8_6.x86_64 #1 SMP Sat Aug 27 02:44:20 EDT 2022

configure options: --build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --program-prefix= --disable-dependency-tracking --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --datadir=/usr/share --includedir=/usr/include --libdir=/usr/lib64 --libexecdir=/usr/libexec --localstatedir=/var --sharedstatedir=/var/lib --mandir=/usr/share/man --infodir=/usr/share/info --disable-libiptc --disable-ipset --enable-snmp --enable-snmp-rfc --enable-sha1 --with-init=systemd build_alias=x86_64-redhat-linux-gnu host_alias=x86_64-redhat-linux-gnu PKG_CONFIG_PATH=:/usr/lib64/pkgconfig:/usr/share/pkgconfig CFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection LDFLAGS=-Wl,-z,relro -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld

Config options:  LVS VRRP VRRP_AUTH OLD_CHKSUM_COMPAT FIB_ROUTING SNMP_V3_FOR_V2 SNMP_VRRP SNMP_CHECKER SNMP_RFCV2 SNMP_RFCV3

System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF IPV6_ADVANCED_API LIBNL3 RTA_ENCAP RTA_EXPIRES RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS FRA_L3MDEV FRA_UID_RANGE RTAX_FASTOPEN_NO_COOKIE RTA_VIA FRA_OIFNAME FRA_PROTOCOL FRA_IP_PROTO FRA_SPORT_RANGE FRA_DPORT_RANGE RTA_TTL_PROPAGATE IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS LWTUNNEL_ENCAP_ILA NET_LINUX_IF_H_COLLISION LIBIPTC_LINUX_NET_IF_H_COLLISION LIBIPVS_NETLINK IPVS_DEST_ATTR_ADDR_FAMILY IPVS_SYNCD_ATTRIBUTES IPVS_64BIT_STATS VRRP_VMAC VRRP_IPVLAN IFLA_LINK_NETNSID CN_PROC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SO_MARK SCHED_RESET_ON_FORK
```

**Distro (please complete the following information):**
 - Name: RHEL
 - Version: 8.6
 - Architecture: x86_64

**Configuration file:**
```
global_defs {
    script_user  nobody
    enable_script_security
    vrrp_startup_delay 10
    max_auto_priority 60
}

vrrp_script check_ssh {
     user nobody
     script "/etc/keepalived/check_ssh.sh 10.1.1.11 22"
     timeout 2
     interval 6
     rise 5
     fall 5
}

vrrp_instance VI_1 {
    state BACKUP
    interface ens192
    virtual_router_id 66
    priority 101
    advert_int 1
    nopreempt

    authentication {
        auth_type PASS
        auth_pass ******
    }
    virtual_ipaddress {
        10.1.1.10
    }
    unicast_peer {
        10.1.1.12
    }
    track_script {
        check_ssh
    }
}

```

**Notify and track scripts**
```
#!/usr/bin/bash
echo quit | timeout 2 /usr/bin/telnet $1 $2 2>/dev/null | grep -q Connected
exit $?
```

**System Log entries**
current master:
```
Oct 12 08:27:43 node1 Keepalived_vrrp[1250]: (VI_1) Master received advert from 10.1.1.12 with same priority 101 but higher IP address than ours
Oct 12 08:27:43 node1 Keepalived_vrrp[1250]: (VI_1) Entering BACKUP STATE
Oct 12 08:27:43 node1 Keepalived_vrrp[1250]: (VI_1) removing VIPs.
```

rebooted node:
```
Oct 12 08:27:40 node2 Keepalived_vrrp[1259]: VRRP_Script(check_ssh) succeeded
Oct 12 08:27:40 node2 Keepalived_vrrp[1259]: (VI_1) Entering BACKUP STATE
Oct 12 08:27:40 node2 Keepalived_vrrp[1259]: VI_1: sending gratuitous ARP for 10.1.1.12
Oct 12 08:27:40 node2 Keepalived_vrrp[1259]: Sending gratuitous ARP on ens192 for 10.1.1.12
Oct 12 08:27:43 node2 Keepalived_vrrp[1259]: (VI_1) Receive advertisement timeout
Oct 12 08:27:43 node2 Keepalived_vrrp[1259]: (VI_1) Entering MASTER STATE
Oct 12 08:27:43 node2 Keepalived_vrrp[1259]: (VI_1) setting VIPs.
Oct 12 08:27:43 node2 Keepalived_vrrp[1259]: (VI_1) Sending/queueing gratuitous ARPs on ens192 for 10.1.1.10
```

**Did keepalived coredump?**
```
no
```

**Additional context**
same symptoms as in https://github.com/acassen/keepalived/issues/2032

setting or further increasing vrrp_startup_delay does not make a difference

the issue seems to be related to some advert timeout:
(VI_1) Receive advertisement timeout (see log above)

how would one debug this?

**Comment by pqarmitage** [comment]
As you have identified (by stopping and restarting keepalived), nopreempt works.

The issue is indeed caused by a timeout due to not receiving an advert. The timeout is 3 * advert_interval + (256 - priority)/256. The timeout is caused by the keepalived process not receiving any adverts, which in turn is caused by the time taken for the network stack to come fully up. Does the keepalived systemd unit file have appropriate After= and Wants= entries?

You ask about how to debug the problem. I suggest you start by increasing `vrrp_startup_delay` to, say, 120 seconds, and see what happens then. From the original master, you could try pinging the backup while it is rebooting to see when it starts responding, but that is not representative of what is happening to keepalived, since replies to ICMP echo requests are handled in the kernel. You could try connecting to a service on the rebooting machine, e.g. ssh, and see when you can first connect. Assuming that your system is connected to a network switch, is it possible that the switch has dropped the MAC address of the system during the reboot, and doesn't learn it again until keepalived sends an advert? It might be worth adding `ExecStartPre=ping SOMEHOST` to the keepalived unit file, to ensure that there is outgoing traffic on the interface before keepalived starts.

If you want to see what is going on in keepalived, you will need to build a version of keepalived having specified `--enable-debug` to configure and then add `--debug-DvEv` to the command line options for keepalived. This will produce a lot of output to syslog, but it will show you what the keepalived vrrp process is doing. It is probably better to also specify `--enable-log-file` to configure, and also add `-g /var/log/keepalived.log` to the keepalived command line options; this will produce several log files, /var/log/keepalived*.log, and the one of interest will be /var/log/keepalived_vrrp.log.

**Comment by wsnipex** [comment]
thank you for your response and excellent advise!
TLDR: firewall issue: the rebooting node was blocking incoming icmp packets until the node made an outgoing icmp request


longer version:
First I verified the systemd service, it had proper After and Wants entries for network-online.target
Following your suggestion I upped the vrrp_startup_delay to 120 sec and verified that the network was fully up and working before keepalived sent the advert. Everything was fine, I could login via ssh, connect to other machines etc, but the problem was still happening, when vrrp started 2min later.
Before resorting to building a debug enabled keepalived, I took a better look at tcpdump on the master node and noticed icmp packets to the rebooted node being dropped. Correlating the timestamps, it sure enough matched with keepalived startup!


checking iptables output gave me this:
```
-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -p icmp -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -p tcp -m state --state NEW -m tcp --dport 22 -j ACCEPT
-A INPUT -p udp -m state --state NEW -m udp --dport 161 -j ACCEPT
-A INPUT -j REJECT --reject-with icmp-host-prohibited
-A FORWARD -j REJECT --reject-with icmp-host-prohibited
```

I'm still not clear why this doesn't work correctly, since there is a general icmp accept rule, but removing this rule fixes it:
`-A INPUT -j REJECT --reject-with icmp-host-prohibited`

**Comment by pqarmitage** [comment]
VRRP has its own protocol number (at the same level as TCP or UDP) - 112.

`-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT`
Once a VRRP packet has been sent, incoming VRRP adverts will be RELATED
`-A INPUT -p icmp -j ACCEPT`
Accept ICMP
`-A INPUT -i lo -j ACCEPT`
Accept anything from the loopback interface
`-A INPUT -p tcp -m state --state NEW -m tcp --dport 22 -j ACCEPT`
Accept incoming TCP port 22 connections - traffic then becomes ESTABLISHED
`-A INPUT -p udp -m state --state NEW -m udp --dport 161 -j ACCEPT`
Accept incoming UDP port 161 packets
`-A INPUT -j REJECT --reject-with icmp-host-prohibited`
Reject anything else inbound with ICMP reason _host-prohibited_ - (this includes protocol 112 - VRRP)
`-A FORWARD -j REJECT --reject-with icmp-host-prohibited`
Reject any packets destined for another system (i.e. routed by this system).

I would suggest that you reinstate the rule
`-A INPUT -j REJECT --reject-with icmp-host-prohibited`
but add a rule BEFORE it:
`-A INPUT -p 112 -j ACCEPT`
which will allow incoming VRRP traffic
